lib/Sema/TypeCheckProtocol.cpp

```diff
@@ -4899,6 +4894,65 @@ void ConformanceChecker::ensureRequirementsAreSatisfied() {
 
 #pragma mark Protocol conformance checking
 
+/// Determine whether mapping the interface type of the given protocol non-type
+/// requirement into the context of the given conformance produces a well-formed
+/// type.
+static bool
+hasInvalidTypeInConformance(const ValueDecl *requirement,
+                            NormalProtocolConformance *conformance) {
+  assert(isa<TypeDecl>(requirement) == false);
+  assert(requirement->getDeclContext()->getSelfProtocolDecl() ==
+         conformance->getProtocol());
+
+  // FIXME: getInterfaceType() on properties returns pseudocontextual types.
+  auto interfaceTy = requirement->getInterfaceType();
+
+  // Skip the 'self' parameter for functions.
+  if (auto *afd = dyn_cast<AbstractFunctionDecl>(requirement))
+    interfaceTy = interfaceTy->castTo<GenericFunctionType>()->getResult();
+
+  // For subscripts, build a regular function type to skip walking generic
+  // requirements.
+  if (auto *gft = interfaceTy->getAs<GenericFunctionType>()) {
+    interfaceTy = FunctionType::get(gft->getParams(), gft->getResult(),
+                                    gft->getExtInfo());
+  }
+
+  if (!interfaceTy->hasTypeParameter())
+    return false;
+
+  const auto subs = SubstitutionMap::getProtocolSubstitutions(
+      conformance->getProtocol(),
+      conformance->getDeclContext()->mapTypeIntoContext(conformance->getType()),
+      ProtocolConformanceRef(conformance));
+
+  class Walker final : public TypeWalker {
+    const SubstitutionMap &Subs;
+
+  public:
+    explicit Walker(const SubstitutionMap &Subs) : Subs(Subs) {}
+
+    Action walkToTypePre(Type ty) override {
+      if (!ty->hasTypeParameter())
+        return Action::SkipChildren;
+
+      auto *const dmt = ty->getAs<DependentMemberType>();
+      if (!dmt)
+        return Action::Continue;
+
+      // We only care about 'Self'-rooted type parameters.
+      if (dmt->getRootGenericParam()->getDepth() != 0)
```

Comment:
dmt->getRootGenericParam()->isEqual(proto->getSelfInterfaceType()) is better

Reply:
Agreed.